src/golang/lib/workflow/operator/base.go

```diff
@@ -279,10 +279,12 @@ func (bo *baseOperator) PersistResult(ctx context.Context) error {
 	)
 
 	for _, outputArtifact := range bo.outputs {
-		// If the operator failed, for now, we assume the downstream artifact is never generated.
-		// In this case, we will mark these artifacts as cancelled.
+		// If the downstream artifact was never generated, we mark it as "cancelled", since the
+		// operator either never ran or did run but hit a user-generated exception.
+		// System-generated errors from things like the type checking of parameters will
+		// still generate downstream artifacts, so those will continue to be marked as "failed".
 		artifactExecState := *execState
-		if execState.Status == shared.FailedExecutionStatus {
+		if !outputArtifact.Computed(ctx) {
 			artifactExecState.Status = shared.CanceledExecutionStatus
```

Comment:
to hide duplicated error messages, we could copy the execState and set `artifactExecState.Error = nil` here

Reply:
What duplicated error messages are we talking about?  We should definitely keep the "Error executing..." msg, but not the other one.

Reply:
oh I see what you mean, hmm

Reply:
answered in the comment below, lemme know what you think